pkg/osquery/runtime/history/instance.go

```diff
@@ -49,14 +52,29 @@ func (i *Instance) Connected(querier Querier) error {
 	i.InstanceId = instanceId
 	i.Version = version
 
+	if err := currentHistory.save(); err != nil {
+		return errors.Wrap(err, "error saving osquery_instance_history")
+	}
+
 	return nil
 }
 
 // InstanceExited sets the exit time and appends provided error (if any) to current osquery instance
-func (i *Instance) Exited(exitError error) {
+func (i *Instance) Exited(exitError error) error {
+	currentHistory.Lock()
+	defer currentHistory.Unlock()
```

Comment:
Why lock here, and not in `currentHistory.save()`?

Reply:
There are [places in the code](https://github.com/kolide/launcher/blob/master/pkg/osquery/runtime/history/history.go#L55) where we manipulate this variable and then save to db. We need to have the lock during this whole process. If we put it only in the save method we would introduce a place where it's getting a lock and then waiting to get the same lock.